implementation/src/main/java/io/smallrye/config/EnvConfigSource.java

```diff
@@ -177,4 +142,141 @@ Object readResolve() {
             return new EnvConfigSource();
         }
     }
+
+    static final class EnvProperty {
+        private final String name;
+
+        public EnvProperty(final String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final EnvProperty that = (EnvProperty) o;
+            return equals(this.name, that.name);
+        }
+
+        @Override
+        public int hashCode() {
+            int h = 0;
+            int length = name.length();
+            if (length >= 2) {
+                if (name.charAt(length - 1) == '_' && name.charAt(length - 2) == '_') {
+                    length = length - 1;
+                }
+            }
+
+            for (int i = 0; i < length; i++) {
+                char c = name.charAt(i);
+                if (i == 0) {
+                    if (c == '%' || c == '_') {
+                        h = 31 * h + 31;
+                        continue;
+                    }
+                }
+
+                if (c == '.' || c == '_' || c == '-' || c == '"' || c == '*' || c == '[' || c == ']' || c == '/') {
```

Comment:
use a `switch` here, cause the JVM can optimize it very effectively

Reply:
Started with only and then kept adding :)